src/iterate-trie-values.ts

```diff
@@ -0,0 +1,239 @@
+const ALWAYS_YIELDS = () => YIELDS_TRUE;
+
+function pushToStack<TValue>(
+	current: Trie<TValue>,
+	visited: Set<Trie<TValue>>,
+	stack: Queue<[Trie<TValue>, number]>,
+	proximity: number,
+) {
+	const { c: sub } = current;
+	for (const k in current.c) {
+		const value = sub[k]!;
+		if (typeof value !== 'string' && !visited.has(value)) {
+			visited.add(current);
+			stack.push([value, proximity]);
+		}
+	}
+}
+
+const identity = (x: unknown) => x;
+function getPrefixTrie(
+	prefixes: string | Iterable<string> | undefined,
+	dataTrie: Trie,
+): [Trie | undefined, number] {
+	if (prefixes === undefined) {
+		return [undefined, 0];
+	}
+	const trie = createEmptyTrie(dataTrie.s);
+	if (typeof prefixes === 'string') {
+		prefixes = [prefixes];
+	}
+	let count = 0;
+	for (const prefix of prefixes) {
+		addWord(trie, prefix);
+		count++;
+	}
+
+	return [trie, count];
+}
+
+function treatOptions<TValue>(
+	prefixOrPrefix: IteratingOptions<TValue>,
+	trie: Trie<TValue>,
+) {
+	const [prefixes, count] = getPrefixTrie(prefixOrPrefix.prefixes, trie);
+	let uniqueness = prefixOrPrefix.uniqueness;
+	const getId = prefixOrPrefix.getId;
+
+	if (prefixes) {
+		const manyPrefixes = count > 1;
+		if (uniqueness === undefined) {
+			uniqueness = manyPrefixes || !getId;
+		} else if (uniqueness === false && manyPrefixes) {
+			throw new TypeError(
+				'It is not allowed to have uniqueness false and more than one prefix',
+			);
+		}
+	}
+	return { prefixes, uniqueness, getId: getId || identity };
+}
+
+function validateIterationParameters<TValue>(
+	prefixOrPrefix: string | IteratingOptions<TValue> | undefined,
+	trie: Trie<TValue>,
+) {
+	let prefixes: Trie | undefined;
+	if (prefixOrPrefix) {
+		if (typeof prefixOrPrefix === 'object') {
+			return treatOptions(prefixOrPrefix, trie);
+		}
+		prefixes = createTrie([prefixOrPrefix], trie.s);
+	}
+
+	return { prefixes, uniqueness: false, getId: identity };
+}
+
+function shouldYieldFactory(
+	uniqueness: boolean | undefined,
+	yielded: Set<unknown>,
+	getId: Function,
+) {
+	if (uniqueness) {
+		return (value: unknown) => {
+			const id = getId(value);
+			if (yielded.has(id)) return { yields: false, id };
+			yielded.add(id);
+			return { yields: true, id };
+		};
+	}
+
+	return ALWAYS_YIELDS;
+}
+
+function* yieldValues<TValue>(
+	shouldYield: ShouldYield,
+	values: TValue[],
+	proximity: number,
+) {
+	const { length } = values;
+	for (let i = 0; i < length; i++) {
+		const value = values[i];
+		const { yields, id } = shouldYield(value);
+		if (yields) {
+			yield { proximity, value, id, count: 1 };
+		}
+	}
+}
+
+function* runIteration<TValue>(
+	trie: Trie<TValue>,
+	uniqueness: boolean | undefined,
+	getId: (x: TValue) => unknown,
+): Iterable<IdItem<TValue>> {
+	const visited = new Set([trie]);
+	const yielded = new Set();
+	const queue = getQueue<[Trie<TValue>, number]>();
+	queue.push([trie, 0]);
+	const shouldYield = shouldYieldFactory(uniqueness, yielded, getId);
+
+	while (queue.hasItems()) {
```

Comment:
What are the advantages of changing the stack to a queue in this scenario? 

*pushToQueue or *addToQueue is a better name

Reply:
I need to do a BFS here to find the more closer node to the starting one